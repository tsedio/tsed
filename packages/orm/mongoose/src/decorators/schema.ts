import {decoratorTypeOf, StoreMerge, useDecorators} from "@tsed/core";
import {registerProvider} from "@tsed/di";
import {Property} from "@tsed/schema";
import {SchemaTypeOptions} from "mongoose";
import {MONGOOSE_SCHEMA} from "../constants";
import {MongooseSchemaOptions} from "../interfaces";
import {getSchema, getSchemaToken} from "../utils/createSchema";

/**
 * Define a class as a Mongoose Schema ready to be used to compose other schemes and models.
 *
 * ### Example
 *
 * ```typescript
 * @MongooseSchema()
 * export class EventSchema {
 *   @Property()
 *   field: string;
 * }
 * ```
 *
 * ### Options
 *
 * - `schemaOptions` (mongoose.SchemaOptions): Option to configure the schema behavior.
 *
 * @param {MongooseSchemaOptions | undefined} options
 * @returns {(target: any) => void}
 * @decorator
 * @mongoose
 * @property
 * @class
 */
export function Schema(options?: MongooseSchemaOptions): (target: any) => void;
export function Schema(definition: SchemaTypeOptions<any>): Function;
export function Schema(options: MongooseSchemaOptions | SchemaTypeOptions<any> = {}) {
  return (...parameters: any[]) => {
    switch (decoratorTypeOf(parameters)) {
      case "property":
        return useDecorators(Property(), StoreMerge(MONGOOSE_SCHEMA, options))(parameters[0], parameters[1], parameters[2]);

      case "class":
        const {token} = getSchemaToken(parameters[0], options);

        registerProvider({
          provide: token,
          deps: [],
          useFactory() {
            return getSchema(parameters[0], options as any);
          }
        });
        break;
    }
  };
}

/**
 * Define a class as a Mongoose Schema ready to be used to compose other schemes and models.
 *
 * ### Example
 *
 * ```typescript
 * @MongooseSchema()
 * export class EventSchema {
 *   @Property()
 *   field: string;
 * }
 * ```
 *
 * ### Options
 *
 * - `schemaOptions` (mongoose.SchemaOptions): Option to configure the schema behavior.
 *
 * @param {MongooseSchemaOptions | undefined} options
 * @returns {(target: any) => void}
 * @decorator
 * @mongoose
 * @property
 * @class
 */
export function MongooseSchema(options?: MongooseSchemaOptions): (target: any) => void;
export function MongooseSchema(definition: SchemaTypeOptions<any>): Function;
export function MongooseSchema(options: MongooseSchemaOptions | SchemaTypeOptions<any> = {}) {
  return Schema(options as any);
}
