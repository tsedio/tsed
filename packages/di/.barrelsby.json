{
  "directory": ["./src/common", "./src/node"],
  "exclude": ["**/__mock__", "**/__mocks__", "**/*.spec.ts"],
  "delete": true
}
