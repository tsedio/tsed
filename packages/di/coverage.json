{
  "statements": 98.52,
  "branches": 92.95,
  "lines": 98.6,
  "functions": 98.11
}
