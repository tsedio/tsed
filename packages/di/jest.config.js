// For a detailed explanation regarding each configuration property, visit:
// https://jestjs.io/docs/en/configuration.html

module.exports = {
  ...require("@tsed/jest-config"),
  coverageThreshold: {
    global: {
      statements: 98.62,
      branches: 96.49,
      lines: 98.62,
      functions: 98.71
    }
  }
};
