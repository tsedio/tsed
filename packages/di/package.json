{
  "name": "@tsed/di",
  "version": "7.70.2",
  "description": "DI module for Ts.ED Framework",
  "source": "./src/index.ts",
  "main": "lib/cjs/index.js",
  "module": "lib/esm/index.js",
  "typings": "./lib/types/index.d.ts",
  "browser": "./lib/browser/di.umd.min.js",
  "exports": {
    "types": "./lib/types/index.d.ts",
    "browser": "./lib/browser/di.umd.min.js",
    "import": "./lib/esm/index.js",
    "require": "./lib/cjs/index.js",
    "default": "./lib/esm/index.js"
  },
  "scripts": {
    "build": "yarn barrels && yarn build:ts && yarn run build:browser",
    "build:browser": "webpack",
    "barrels": "barrels",
    "test": "jest --max-workers=2 && jest-coverage-thresholds-bumper",
    "build:ts": "tsc --build tsconfig.json && tsc --build tsconfig.esm.json"
  },
  "dependencies": {
    "tslib": "2.6.2"
  },
  "devDependencies": {
    "@tsed/barrels": "workspace:*",
    "@tsed/core": "workspace:*",
    "@tsed/logger": ">=6.7.5",
    "@tsed/schema": "workspace:*",
    "@tsed/typescript": "workspace:*",
    "cross-env": "7.0.3",
    "eslint": "^8.57.0",
    "jest": "^29.7.0"
  },
  "peerDependencies": {
    "@tsed/core": "7.70.2",
    "@tsed/logger": ">=6.7.5",
    "@tsed/schema": "7.70.2"
  },
  "peerDependenciesMeta": {
    "@tsed/core": {
      "optional": false
    },
    "@tsed/logger": {
      "optional": false
    },
    "@tsed/schema": {
      "optional": false
    }
  }
}
