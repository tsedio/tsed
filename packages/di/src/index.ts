/**
 * @file Automatically generated by barrelsby.
 */

export * from "./constants/constants";
export * from "./decorators/configuration";
export * from "./decorators/constant";
export * from "./decorators/controller";
export * from "./decorators/inject";
export * from "./decorators/injectable";
export * from "./decorators/intercept";
export * from "./decorators/interceptor";
export * from "./decorators/lazyInject";
export * from "./decorators/module";
export * from "./decorators/opts";
export * from "./decorators/overrideProvider";
export * from "./decorators/scope";
export * from "./decorators/service";
export * from "./decorators/useOpts";
export * from "./decorators/value";
export * from "./domain/Container";
export * from "./domain/ContextLogger";
export * from "./domain/DIContext";
export * from "./domain/InjectablePropertyType";
export * from "./domain/LocalsContainer";
export * from "./domain/Provider";
export * from "./domain/ProviderScope";
export * from "./domain/ProviderType";
export * from "./errors/InjectionError";
export * from "./errors/UndefinedTokenError";
export * from "./interfaces/DIConfigurationOptions";
export * from "./interfaces/DILogger";
export * from "./interfaces/DIResolver";
export * from "./interfaces/ProviderOpts";
export * from "./interfaces/InjectableProperties";
export * from "./interfaces/InterceptorContext";
export * from "./interfaces/InterceptorMethods";
export * from "./interfaces/InvokeOptions";
export * from "./interfaces/OnDestroy";
export * from "./interfaces/OnInit";
export * from "./interfaces/RegistrySettings";
export * from "./interfaces/ResolvedInvokeOptions";
export * from "./interfaces/TokenProvider";
export * from "./registries/GlobalProviders";
export * from "./registries/ProviderRegistry";
export * from "./services/DIConfiguration";
export * from "./services/DILogger";
export * from "./services/DITest";
export * from "./services/InjectorService";
export * from "./utils/colors";
export * from "./utils/createContainer";
export * from "./utils/getConfiguration";
export * from "./utils/setLoggerLevel";
