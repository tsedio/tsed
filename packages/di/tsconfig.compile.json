{
  "extends": "../../tsconfig.compile.json",
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "lib/cjs"
  },
  "exclude": [
    "node_modules",
    "test",
    "lib",
    "benchmark",
    "coverage",
    "spec",
    "**/*.benchmark.ts",
    "**/*.spec.ts",
    "keys",
    "jest.config.js",
    "**/__mock__/**"
  ]
}
