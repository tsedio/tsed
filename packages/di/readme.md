<p style="text-align: center" align="center">
 <a href="https://tsed.io" target="_blank"><img src="https://tsed.io/tsed-og.png" width="200" alt="Ts.ED logo"/></a>
</p>

<div align="center">
   <h1>@tsed/di</h1>

[![Build & Release](https://github.com/tsedio/tsed/workflows/Build%20&%20Release/badge.svg)](https://github.com/tsedio/tsed/actions?query=workflow%3A%22Build+%26+Release%22)
[![PR Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](https://github.com/tsedio/tsed/blob/master/CONTRIBUTING.md)
[![Coverage Status](https://coveralls.io/repos/github/tsedio/tsed/badge.svg?branch=production)](https://coveralls.io/github/tsedio/tsed?branch=production)
[![npm version](https://badge.fury.io/js/%40tsed%2Fcommon.svg)](https://badge.fury.io/js/%40tsed%2Fcommon)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![backers](https://opencollective.com/tsed/tiers/badge.svg)](https://opencollective.com/tsed)

</div>

<div align="center">
  <a href="https://tsed.io/">Website</a>
  <span>&nbsp;&nbsp;‚Ä¢&nbsp;&nbsp;</span>
  <a href="https://tsed.io/getting-started/">Getting started</a>
  <span>&nbsp;&nbsp;‚Ä¢&nbsp;&nbsp;</span>
  <a href="https://api.tsed.io/rest/slack/tsedio/tsed">Slack</a>
  <span>&nbsp;&nbsp;‚Ä¢&nbsp;&nbsp;</span>
  <a href="https://twitter.com/TsED_io">Twitter</a>
</div>

<hr />

A package of Ts.ED framework. See website: https://tsed.io/

# Installation

You can get the latest release and the type definitions using npm:

```bash
npm install --save @tsed/di
```

> **Important!** TsExpressDecorators requires Node >= 6, Express >= 4, TypeScript >= 2.0 and
> the `experimentalDecorators`, `emitDecoratorMetadata`, `types` and `lib` compilation
> options in your `tsconfig.json` file.

```json
{
  "compilerOptions": {
    "target": "es2016",
    "lib": ["es2016"],
    "typeRoots": ["./node_modules/@types"],
    "module": "commonjs",
    "moduleResolution": "node",
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "allowSyntheticDefaultImports": true
  },
  "exclude": ["node_modules"]
}
```

## Introduction

Basically, almost everything may be considered as a provider ‚Äì service, factory, intereceptors, and so on.
All of them can inject dependencies, meaning, they can create various relationships with each other.
But in fact, a provider is nothing else than just a simple class annotated with an `@Injectable()` decorator.

## Usage

Here is a basic usage to declare an injectable service to another one:

```typescript
import {Injectable} from "@tsed/di";
import {Calendar} from "./models/calendar";

@Injectable()
export class CalendarsService {
  private readonly calendars: Calendar[] = [];

  create(calendar: Calendar) {
    this.calendars.push(calendar);
  }

  findAll(): Calendar[] {
    return this.calendars;
  }
}
```

Here's a CalendarsService, a basic class with one property and two methods. The only new trait is that it uses the `@Injectable()` decorator.
The `@Injectable()` attaches the metadata, thereby Ts.ED knows that this class is a provider.

Now we have the service class already done, let's use it inside a controller:

```typescript
import {Controller, Post, Body, Get} from "@tsed/common";
import {CalendarsService} from "./CalendarsService";
import {Calendar} from "./models/Calendar";

@Controller("/calendars")
export class CalendarCtrl {
  constructor(private readonly calendarsService: CalendarsService) {}

  @Post()
  async create(@Body() calendar: Calendar) {
    this.calendarsService.create(calendar);
  }

  @Get()
  async findAll(): Promise<Calendar[]> {
    return this.calendarsService.findAll();
  }
}
```

> Note: Controller isn't a part of `@tsed/di`. `@Controller` decorator is exposed by `@tsed/common` package because it's a specific provider
> used by the Ts.ED framework. Ts.ED DI allow you to define your own Provider and decorator.

Finally, we can load the injector and use:

```typescript
import {InjectorService} from "@tsed/di";
import {CalendarCtrl} from "./CalendarCtrl";

async function bootstrap() {
  const injector = new InjectorService();

  // Load all providers registered via @Injectable decorator
  await injector.load();

  const calendarController = injector.get<CalendarCtrl>();

  await calendarController.create(new Calendar());

  // And finally destroy injector and his instances (see injector hooks)
  await injector.destroy();
}

bootstrap();
```

## Custom providers

To organize your code Ts.ED DI provide different kind of providers:

- Provider can be declared with `@Injectable`,
- Service can be declared with `@Service`,
- Interceptor can be declared with `@Interceptor`,
- Factory and Value can be declared with `registerFactory` and `registerValue`.

See more details on our documentation https://tsed.io/providers.html

## Contributors

Please read [contributing guidelines here](https://tsed.io/CONTRIBUTING.html)

<a href="https://github.com/tsedio/ts-express-decorators/graphs/contributors"><img src="https://opencollective.com/tsed/contributors.svg?width=890" /></a>

## Backers

Thank you to all our backers! üôè [[Become a backer](https://opencollective.com/tsed#backer)]

<a href="https://opencollective.com/tsed#backers" target="_blank"><img src="https://opencollective.com/tsed/tiers/backer.svg?width=890"></a>

## Sponsors

Support this project by becoming a sponsor. Your logo will show up here with a link to your website. [[Become a sponsor](https://opencollective.com/tsed#sponsor)]

## License

The MIT License (MIT)

Copyright (c) 2016 - 2018 Romain Lenzotti

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
