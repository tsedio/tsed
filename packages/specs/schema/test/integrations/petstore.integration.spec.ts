import {Controller} from "@tsed/di";
import {Use} from "@tsed/platform-middlewares";
import {BodyParams, PathParams} from "@tsed/platform-params";
import {CollectionOf, Name, Property} from "@tsed/schema";
import {
  Delete,
  Description,
  Enum,
  Example,
  Get,
  getSpec,
  Groups,
  Partial,
  Patch,
  Post,
  Publish,
  Put,
  Required,
  Returns,
  SpecTypes,
  Subscribe
} from "../../src/index.js";

class PetCategory {
  @Required()
  @Groups("!partial", "!create", "!update")
  id: string;

  @Required()
  @Example("doggie")
  name: string;
}

enum PetStatus {
  AVAILABLE = "available",
  PENDING = "pending",
  SOLD = "sold"
}

class Pet {
  @Required()
  @Groups("!partial", "!create", "!update")
  id: string;

  @Required()
  @Example("doggie")
  name: string;

  @Property()
  category: PetCategory;

  @CollectionOf(String)
  tags: string[];

  @Enum(PetStatus)
  status: PetStatus;
}

@Controller("/")
@Name("PetStore")
class PetController {
  @Use("/")
  middleware(@PathParams("id") id: string) {}

  @Get("/:id")
  @Publish("pet.get")
  @Subscribe("pet.get")
  @Returns(200, Pet).Description("Returns a pet")
  @Returns(404)
  get(@PathParams("id") id: string) {
    return null;
  }

  @Get("/")
  @Publish("pet.getAll")
  @Subscribe("pet.getAll")
  @Returns(200, Array).Of(Pet).Description("Returns all pets")
  getAll() {
    return [];
  }

  @Patch("/:id")
  @Publish("pet.patch")
  @Subscribe("pet.updated")
  @Returns(200, Pet).Description("Returns a pet")
  @Returns(404)
  @Description("Patch a pet")
  patch(@PathParams("id") id: string, @BodyParams() @Partial() partial: Pet) {
    return null;
  }

  @Post("/:id")
  @Publish("pet.update")
  @Subscribe("pet.created")
  @Returns(200, Pet).Description("Returns a pet")
  @Returns(404)
  @Description("Update a pet")
  post(@BodyParams() @Groups("update") pet: Pet) {
    return null;
  }

  @Put("/")
  @Publish("pet.create")
  @Subscribe("pet.updated")
  @Returns(201, Pet)
  @Returns(404)
  @Description("Create a pet")
  put(@BodyParams() @Groups("create") pet: Pet) {
    return null;
  }

  @Delete("/:id")
  @Publish("pet.delete")
  @Subscribe("pet.deleted")
  @Returns(204)
  @Returns(404)
  @Description("Delete a pet")
  delete(@PathParams("id") id: string) {
    return null;
  }
}

describe("PetStore", () => {
  describe("OpenSpec", () => {
    it("should generate the spec", () => {
      const spec = getSpec(PetController, {specType: SpecTypes.OPENAPI});

      expect(spec).toMatchSnapshot();
    });
  });

  describe("AsyncAPI", () => {
    it("should generate the spec", () => {
      const spec = getSpec(PetController, {specType: SpecTypes.ASYNCAPI});

      expect(spec).toMatchSnapshot();
    });
  });
});
