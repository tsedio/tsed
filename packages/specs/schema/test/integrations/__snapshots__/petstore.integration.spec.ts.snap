// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PetStore AsyncAPI should generate the spec 1`] = `
Object {
  "channels": Object {
    "/": Object {
      "publish": Object {
        "message": Object {
          "oneOf": Array [
            Object {
              "$ref": "#/components/messages/pet.get",
            },
            Object {
              "$ref": "#/components/messages/pet.getAll",
            },
            Object {
              "$ref": "#/components/messages/pet.patch",
            },
            Object {
              "$ref": "#/components/messages/pet.update",
            },
            Object {
              "$ref": "#/components/messages/pet.create",
            },
            Object {
              "$ref": "#/components/messages/pet.delete",
            },
          ],
        },
        "operationId": "publishPetStore",
      },
      "subscribe": Object {
        "message": Object {
          "oneOf": Array [
            Object {
              "$ref": "#/components/messages/pet.get",
            },
            Object {
              "$ref": "#/components/messages/pet.getAll",
            },
            Object {
              "$ref": "#/components/messages/pet.updated",
            },
            Object {
              "$ref": "#/components/messages/pet.created",
            },
            Object {
              "$ref": "#/components/messages/pet.updated",
            },
            Object {
              "$ref": "#/components/messages/pet.deleted",
            },
          ],
        },
        "operationId": "subscribePetStore",
      },
    },
  },
  "components": Object {
    "messages": Object {
      "pet.create": Object {
        "description": "Create a pet",
        "payload": Object {
          "$ref": "#/components/schemas/PetCreate",
        },
        "x-response": Object {
          "$ref": "#/components/messages/pet.updated",
        },
      },
      "pet.created": Object {
        "description": "Update a pet",
        "x-response": Object {
          "$ref": "#/components/schemas/PetCreatedSubscribeResponse",
        },
      },
      "pet.delete": Object {
        "description": "Delete a pet",
        "payload": Object {
          "properties": Object {
            "id": Object {
              "type": "string",
            },
          },
          "type": "object",
        },
        "x-response": Object {
          "$ref": "#/components/messages/pet.deleted",
        },
      },
      "pet.deleted": Object {
        "description": "Delete a pet",
        "x-response": Object {
          "$ref": "#/components/schemas/PetDeletedSubscribeResponse",
        },
      },
      "pet.get": Object {
        "payload": Object {
          "properties": Object {
            "id": Object {
              "type": "string",
            },
          },
          "type": "object",
        },
        "x-response": Object {
          "$ref": "#/components/messages/pet.get",
        },
      },
      "pet.getAll": Object {
        "x-response": Object {
          "$ref": "#/components/messages/pet.getAll",
        },
      },
      "pet.patch": Object {
        "description": "Patch a pet",
        "payload": Object {
          "allOf": Array [
            Object {
              "properties": Object {
                "id": Object {
                  "type": "string",
                },
              },
              "type": "object",
            },
            Object {
              "$ref": "#/components/schemas/PetPartial",
            },
          ],
        },
        "x-response": Object {
          "$ref": "#/components/messages/pet.updated",
        },
      },
      "pet.update": Object {
        "description": "Update a pet",
        "payload": Object {
          "$ref": "#/components/schemas/PetUpdate",
        },
        "x-response": Object {
          "$ref": "#/components/messages/pet.created",
        },
      },
      "pet.updated": Object {
        "description": "Create a pet",
        "x-response": Object {
          "$ref": "#/components/schemas/PetUpdatedSubscribeResponse",
        },
      },
    },
    "schemas": Object {
      "Pet": Object {
        "properties": Object {
          "category": Object {
            "$ref": "#/components/schemas/PetCategory",
          },
          "id": Object {
            "minLength": 1,
            "type": "string",
          },
          "name": Object {
            "example": "doggie",
            "minLength": 1,
            "type": "string",
          },
          "status": Object {
            "enum": Array [
              "available",
              "pending",
              "sold",
            ],
            "type": "string",
          },
          "tags": Object {
            "items": Object {
              "type": "string",
            },
            "type": "array",
          },
        },
        "required": Array [
          "id",
          "name",
        ],
        "type": "object",
      },
      "PetCategory": Object {
        "properties": Object {
          "id": Object {
            "minLength": 1,
            "type": "string",
          },
          "name": Object {
            "example": "doggie",
            "minLength": 1,
            "type": "string",
          },
        },
        "required": Array [
          "id",
          "name",
        ],
        "type": "object",
      },
      "PetCreate": Object {
        "properties": Object {
          "category": Object {
            "$ref": "#/components/schemas/PetCategory",
          },
          "name": Object {
            "example": "doggie",
            "minLength": 1,
            "type": "string",
          },
          "status": Object {
            "enum": Array [
              "available",
              "pending",
              "sold",
            ],
            "type": "string",
          },
          "tags": Object {
            "items": Object {
              "type": "string",
            },
            "type": "array",
          },
        },
        "required": Array [
          "name",
        ],
        "type": "object",
      },
      "PetCreatePublishPayload": Object {},
      "PetCreatedSubscribeResponse": Object {
        "properties": Object {
          "data": Object {
            "oneOf": Array [
              Object {
                "schema": Object {
                  "type": "object",
                },
              },
              Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Pet",
                },
              },
            ],
          },
          "status": Object {
            "enum": Array [
              404,
              200,
            ],
            "type": "number",
          },
          "statusText": Object {
            "enum": Array [
              "Not Found",
              "Returns a pet",
            ],
            "type": "string",
          },
        },
        "required": Array [
          "status",
        ],
        "type": "object",
      },
      "PetDeletePublishPayload": Object {},
      "PetDeletedSubscribeResponse": Object {
        "properties": Object {
          "data": Object {
            "schema": Object {
              "type": "object",
            },
          },
          "status": Object {
            "enum": Array [
              404,
              204,
            ],
            "type": "number",
          },
          "statusText": Object {
            "enum": Array [
              "Not Found",
              "No Content",
            ],
            "type": "string",
          },
        },
        "required": Array [
          "status",
        ],
        "type": "object",
      },
      "PetGetAllPublishPayload": Object {},
      "PetGetAllSubscribeResponse": Object {
        "properties": Object {
          "data": Object {
            "schema": Object {
              "items": Object {
                "$ref": "#/components/schemas/Pet",
              },
              "type": "array",
            },
          },
          "status": Object {
            "enum": Array [
              200,
            ],
            "type": "number",
          },
          "statusText": Object {
            "enum": Array [
              "Returns all pets",
            ],
            "type": "string",
          },
        },
        "required": Array [
          "status",
        ],
        "type": "object",
      },
      "PetGetPublishPayload": Object {},
      "PetGetSubscribeResponse": Object {
        "properties": Object {
          "data": Object {
            "oneOf": Array [
              Object {
                "schema": Object {
                  "type": "object",
                },
              },
              Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Pet",
                },
              },
            ],
          },
          "status": Object {
            "enum": Array [
              404,
              200,
            ],
            "type": "number",
          },
          "statusText": Object {
            "enum": Array [
              "Not Found",
              "Returns a pet",
            ],
            "type": "string",
          },
        },
        "required": Array [
          "status",
        ],
        "type": "object",
      },
      "PetPartial": Object {
        "properties": Object {
          "category": Object {
            "$ref": "#/components/schemas/PetCategory",
          },
          "name": Object {
            "example": "doggie",
            "type": "string",
          },
          "status": Object {
            "enum": Array [
              "available",
              "pending",
              "sold",
            ],
            "type": "string",
          },
          "tags": Object {
            "items": Object {
              "type": "string",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "PetPatchPublishPayload": Object {},
      "PetUpdate": Object {
        "properties": Object {
          "category": Object {
            "$ref": "#/components/schemas/PetCategory",
          },
          "name": Object {
            "example": "doggie",
            "minLength": 1,
            "type": "string",
          },
          "status": Object {
            "enum": Array [
              "available",
              "pending",
              "sold",
            ],
            "type": "string",
          },
          "tags": Object {
            "items": Object {
              "type": "string",
            },
            "type": "array",
          },
        },
        "required": Array [
          "name",
        ],
        "type": "object",
      },
      "PetUpdatePublishPayload": Object {},
      "PetUpdatedSubscribeResponse": Object {
        "properties": Object {
          "data": Object {
            "oneOf": Array [
              Object {
                "schema": Object {
                  "type": "object",
                },
              },
              Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Pet",
                },
              },
            ],
          },
          "status": Object {
            "enum": Array [
              404,
              201,
            ],
            "type": "number",
          },
          "statusText": Object {
            "enum": Array [
              "Not Found",
              "Created",
            ],
            "type": "string",
          },
        },
        "required": Array [
          "status",
        ],
        "type": "object",
      },
    },
  },
  "tags": Array [],
}
`;

exports[`PetStore OpenSpec should generate the spec 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "Pet": Object {
        "properties": Object {
          "category": Object {
            "$ref": "#/components/schemas/PetCategory",
          },
          "id": Object {
            "minLength": 1,
            "type": "string",
          },
          "name": Object {
            "example": "doggie",
            "minLength": 1,
            "type": "string",
          },
          "status": Object {
            "enum": Array [
              "available",
              "pending",
              "sold",
            ],
            "type": "string",
          },
          "tags": Object {
            "items": Object {
              "type": "string",
            },
            "type": "array",
          },
        },
        "required": Array [
          "id",
          "name",
        ],
        "type": "object",
      },
      "PetCategory": Object {
        "properties": Object {
          "id": Object {
            "minLength": 1,
            "type": "string",
          },
          "name": Object {
            "example": "doggie",
            "minLength": 1,
            "type": "string",
          },
        },
        "required": Array [
          "id",
          "name",
        ],
        "type": "object",
      },
      "PetCreate": Object {
        "properties": Object {
          "category": Object {
            "$ref": "#/components/schemas/PetCategory",
          },
          "name": Object {
            "example": "doggie",
            "minLength": 1,
            "type": "string",
          },
          "status": Object {
            "enum": Array [
              "available",
              "pending",
              "sold",
            ],
            "type": "string",
          },
          "tags": Object {
            "items": Object {
              "type": "string",
            },
            "type": "array",
          },
        },
        "required": Array [
          "name",
        ],
        "type": "object",
      },
      "PetPartial": Object {
        "properties": Object {
          "category": Object {
            "$ref": "#/components/schemas/PetCategory",
          },
          "name": Object {
            "example": "doggie",
            "type": "string",
          },
          "status": Object {
            "enum": Array [
              "available",
              "pending",
              "sold",
            ],
            "type": "string",
          },
          "tags": Object {
            "items": Object {
              "type": "string",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "PetUpdate": Object {
        "properties": Object {
          "category": Object {
            "$ref": "#/components/schemas/PetCategory",
          },
          "name": Object {
            "example": "doggie",
            "minLength": 1,
            "type": "string",
          },
          "status": Object {
            "enum": Array [
              "available",
              "pending",
              "sold",
            ],
            "type": "string",
          },
          "tags": Object {
            "items": Object {
              "type": "string",
            },
            "type": "array",
          },
        },
        "required": Array [
          "name",
        ],
        "type": "object",
      },
    },
  },
  "paths": Object {
    "/": Object {
      "get": Object {
        "operationId": "petStoreGetAll",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "items": Object {
                    "$ref": "#/components/schemas/Pet",
                  },
                  "type": "array",
                },
              },
            },
            "description": "Returns all pets",
          },
        },
        "tags": Array [
          "PetStore",
        ],
      },
      "put": Object {
        "description": "Create a pet",
        "operationId": "petStorePut",
        "parameters": Array [],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "$ref": "#/components/schemas/PetCreate",
              },
            },
          },
          "required": false,
        },
        "responses": Object {
          "201": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Pet",
                },
              },
            },
            "description": "Created",
          },
          "404": Object {
            "content": Object {
              "*/*": Object {
                "schema": Object {
                  "type": "object",
                },
              },
            },
            "description": "Not Found",
          },
        },
        "tags": Array [
          "PetStore",
        ],
      },
    },
    "/{id}": Object {
      "delete": Object {
        "description": "Delete a pet",
        "operationId": "petStoreDelete",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {
          "204": Object {
            "description": "No Content",
          },
          "404": Object {
            "content": Object {
              "*/*": Object {
                "schema": Object {
                  "type": "object",
                },
              },
            },
            "description": "Not Found",
          },
        },
        "tags": Array [
          "PetStore",
        ],
      },
      "get": Object {
        "operationId": "petStoreGet",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Pet",
                },
              },
            },
            "description": "Returns a pet",
          },
          "404": Object {
            "content": Object {
              "*/*": Object {
                "schema": Object {
                  "type": "object",
                },
              },
            },
            "description": "Not Found",
          },
        },
        "tags": Array [
          "PetStore",
        ],
      },
      "patch": Object {
        "description": "Patch a pet",
        "operationId": "petStorePatch",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "$ref": "#/components/schemas/PetPartial",
              },
            },
          },
          "required": false,
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Pet",
                },
              },
            },
            "description": "Returns a pet",
          },
          "404": Object {
            "content": Object {
              "*/*": Object {
                "schema": Object {
                  "type": "object",
                },
              },
            },
            "description": "Not Found",
          },
        },
        "tags": Array [
          "PetStore",
        ],
      },
      "post": Object {
        "description": "Update a pet",
        "operationId": "petStorePost",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "$ref": "#/components/schemas/PetUpdate",
              },
            },
          },
          "required": false,
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Pet",
                },
              },
            },
            "description": "Returns a pet",
          },
          "404": Object {
            "content": Object {
              "*/*": Object {
                "schema": Object {
                  "type": "object",
                },
              },
            },
            "description": "Not Found",
          },
        },
        "tags": Array [
          "PetStore",
        ],
      },
    },
  },
  "tags": Array [
    Object {
      "name": "PetStore",
    },
  ],
}
`;
